docs/decisions/0035-embedding-abstraction.md

```diff
@@ -0,0 +1,257 @@
+
+```csharp
+public class EmbeddingContent<TEmbedding> : KernelContent where TEmbedding : unmanaged
+{
+    public EmbeddingContent(
+        IReadOnlyList<ReadOnlyMemory<TEmbedding>> data,
+        string? modelId = null,
+        object? innerContent = null,
+        IReadOnlyDictionary<string, object?>? metadata = null)
+        : base(innerContent, modelId, metadata)
+    {
+        this.Data = data;
+    }
+
+    public IReadOnlyList<ReadOnlyMemory<TEmbedding>> Data { get; set; }
+}
+
+public interface ITextEmbeddingGenerationService : IAIService
+{
+    Task<IReadOnlyList<EmbeddingContent<double>>> GenerateEmbeddingsAsync(
+        IList<string> data,
+        Kernel? kernel = null,
+        PromptExecutionSettings? executionSettings = null,
+        CancellationToken cancellationToken = default);
+}
+
+public interface IImageEmbeddingGenerationService : IAIService
+{
+    Task<IReadOnlyList<EmbeddingContent<double>>> GenerateEmbeddingsAsync(
+        IList<ImageContent> data,
+        Kernel? kernel = null,
+        PromptExecutionSettings? executionSettings = null,
+        CancellationToken cancellationToken = default);
+}
+```
+Pros:
+- Allows for generating embeddings from different types of data.
+- Allows for parameterizing the query.
+- Allows for returning metadata.
+- Interface segregation.
+
+Cons:
+- No generic interface.
+
+### Option 3 [Proposed] - Common interface for embeddings generation with metadata and query parameterization
+
+Similar to option 2, but with a common interface for generating embeddings from different types of data.
+
+```csharp
+public class EmbeddingContent<TEmbedding> : KernelContent where TEmbedding : unmanaged
+{
+    public EmbeddingContent(
+        IReadOnlyList<ReadOnlyMemory<TEmbedding>> data,
+        string? modelId = null,
+        object? innerContent = null,
+        IReadOnlyDictionary<string, object?>? metadata = null)
+        : base(innerContent, modelId, metadata)
+    {
+        this.Data = data;
+    }
+
+    public IReadOnlyList<ReadOnlyMemory<TEmbedding>> Data { get; set; }
+}
+
+public interface IEmbeddingGenerationService : IAIService
+{
+    Task<IReadOnlyList<EmbeddingContent<double>>> GenerateEmbeddingsAsync(
+        IList<string> data,
+        Kernel? kernel = null,
+        PromptExecutionSettings? executionSettings = null,
+        CancellationToken cancellationToken = default);
+    
+    Task<IReadOnlyList<EmbeddingContent<double>>> GenerateEmbeddingsAsync(
+        IList<ImageContent> data,
+        Kernel? kernel = null,
+        PromptExecutionSettings? executionSettings = null,
+        CancellationToken cancellationToken = default);
+}
+```
+
+Pros:
+- Allows for generating embeddings from different types of data.
+- Allows for parameterizing the query.
+- Allows for returning metadata.
+
+Cons:
+- No generic interface.
+- One common interface for different types of data. Many models don't support all types of data.
```

Comment:
That's correct, it would be nice to avoid this approach since connectors will be forced to implement functionality which they don't support. But it's still an option, so it's good that you mentioned it in ADR, thanks!